src/slice.c

```diff
@@ -559,6 +559,14 @@ static SEXP chr_as_index(SEXP i, SEXP names) {
 }
 
 SEXP vec_as_index(SEXP i, R_len_t n, SEXP names) {
+  if (OBJECT(i)) {
+    Rf_errorcall(R_NilValue, "`i` must be an integer, character, or logical vector, not an object.");
```

Comment:
Probably better to use `vec_cast()` to integer? Or `vec_coerce_cast()`, so that only coercible types are allowed (e.g. not `c("1", "2")`).